core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala

```diff
@@ -667,26 +667,42 @@ class ControllerIntegrationTest extends ZooKeeperTestHarness {
 
     TestUtils.createTopic(zkClient, topic, partitionReplicaAssignment = expectedReplicaAssignment, servers = servers)
 
+    def waitUntilIsrSizeConditionHolds(topic: String, partition: Int, servers: Seq[KafkaServer],
+                                       isrSizePredicate: Int => Boolean, msg: String): Unit = {
+      TestUtils.waitUntilTrue(
+        () => servers.forall(
+          _.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic, partition) match {
+            case Some(partitionInfo) => isrSizePredicate(partitionInfo.isr.size)
+            case None => false  // If partitionInfo/ISR is not ready yet, the condition cannot be evaluated
+          }
+        ),
+        msg
+      )
+    }
+
+    def equalsTo(i: Int): Int => Boolean =  _ == i
+    def notEqualsTo(i: Int): Int => Boolean =  _ != i
+
     // Attempt to shut down and then restart broker1
     val broker1 = servers.filter(s => s.config.brokerId == 1).head
     broker1.shutdown()
     var activeServers = servers.filter(s => s.config.brokerId != 1)
-    TestUtils.waitUntilTrue(() =>
-      activeServers.forall(_.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic,partition).get.isr.size != 2),
-      "Topic test not created after timeout")
+    waitUntilIsrSizeConditionHolds(topic, partition, activeServers,
+                                   isrSizePredicate = notEqualsTo(2),
+                                   "Topic test not created or metadata not propagated after timeout")
 
     broker1.startup()
-    TestUtils.waitUntilTrue(() =>
-      servers.forall(_.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic, partition).get.isr.size == 2),
-      "The ISR does not include the full set after the offline broker is restarted")
+    waitUntilIsrSizeConditionHolds(topic, partition, servers,
+                                   isrSizePredicate = equalsTo(2),
```

Comment:
Probably replace 
isrSizePredicate = equalsTo(2)
with
_ == 2
considering the latter is shorter and seems easier for understanding.

Reply:
Fixed

Reply:
Fixed